src/snowflake/snowpark/_internal/ast.py

```diff
@@ -20,13 +28,99 @@ def expr_to_dataframe_expr(expr):
+    """
+    Decodes Snowpark REST response to protobuf response message.
+    Args:
+        res: Dictionary representing a snowpark REST response.
+
+    Returns:
+        Protobuf response message.
+    """
+
+    # Check if response resulted in error code, if so decode.
+    data = res["data"]
+
+    # Similar to existing Snowpark client, surface errors as SnowparkSQLException.
+    if "errorCode" in data.keys():
+        is_internal_error = data["internalError"]
+        error_code = data["errorCode"]
+        sfqid = data["queryId"]
+        error_message = res["message"]
+
+        if is_internal_error:
+            error_message = "INTERNAL ERROR: " + error_message
+
+        raise SnowparkSQLException(error_message, error_code=error_code, sfqid=sfqid)
+
+    # `data` is given as b64 encoded rowset result.
+    # First retrieve response from Snowpark Python connector format,
+    # then convert to IR compatible protobuf message.
+
+    if data["queryResultFormat"] == "arrow" and "rowsetBase64" in data.keys():
+        # This code is a stripped down version from the Snowflake Python connector.
+        # The response object of the IR is delivered as a single STRING column at the moment.
+        # We may change this in the near future.
+        rowset_b64 = data["rowsetBase64"]
+
+        total_len: int = data.get("total", 0)
+        first_chunk_len = total_len
+        arrow_context = ArrowConverterContext(session_parameters)
+
+        schema: Sequence[ResultMetadataV2] = [
+            ResultMetadataV2.from_column(col) for col in data["rowtype"]
+        ]
+
+        if "chunks" in data:
+            raise NotImplementedError("decoding chunks not yet supported")
+
+        first_chunk = ArrowResultBatch.from_data(
+            rowset_b64,
+            first_chunk_len,
+            arrow_context,
+            True,
+            True,  # does not matter
+            schema,
+            True,  # does not matter
+        )
+
+        assert (
+            first_chunk.rowcount == 1
+        ), "Result should consist of single row holding protobuf response"
+
+        dict_result = first_chunk.to_arrow().to_pydict()
+
+        # should be single key, value pair:
+        response_as_json: str = list(dict_result.values())[0][0]
+
+        response_as_dict = json.loads(response_as_json)
+
+        if response_as_dict["status"] != 200:
+            raise SnowparkClientExceptionMessages.IR_MESSAGE(
+                f"Coprocessor returned status {response_as_dict['status']}"
+            )
+
+        # Should be also able to load the json directly into a python dict via json.loads(...),
+        # however map here to protobuf to make sure contents align with protobuf message.
+        response = ParseDict(response_as_dict["data"], proto.Response())
+        return response
+    else:
+        raise NotImplementedError("Only inline arrow result decode supported yet.")
```

Comment:
...supported at the moment (or currently). "Yet" doesn't really parse correctly here.